create role eos_validate_state

ksator opened the pull request:
issue #206  
Create a new role, to validate EOS devices state     
It consumes the same input as the role eos_cli_config_gen    
The readme file of this role provides the instructions

ksator committed:
create role eos_validate_state ###206

ksator committed:
fix tasks on eos_validate_state role to skip empty dictionnary #206

carlbuchmann committed:
fix lint issues + update structure for AVD std.

carlbuchmann committed:
add reference to eos_validate_state role

carlbuchmann committed:
remove whitespace + formating

carlbuchmann reviewed (commented):
Tests completed:
- [x] vEOS-LAB AVD Example Topology
- [x] Physical Topology

vEOS-LAB test results:
542 tests completed successfully across 14 devices!

```
PLAY RECAP ******************************************************************************************************************************************************************
DC1-BL1A                   : ok=12   changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
DC1-BL1B                   : ok=12   changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
DC1-L2LEAF1A               : ok=9    changed=0    unreachable=0    failed=0    skipped=3    rescued=0    ignored=0   
DC1-L2LEAF2A               : ok=10   changed=0    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   
DC1-L2LEAF2B               : ok=10   changed=0    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   
DC1-LEAF1A                 : ok=11   changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
DC1-LEAF2A                 : ok=12   changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
DC1-LEAF2B                 : ok=12   changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
DC1-SPINE1                 : ok=11   changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
DC1-SPINE2                 : ok=11   changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
DC1-SPINE3                 : ok=11   changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
DC1-SPINE4                 : ok=11   changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
DC1-SVC3A                  : ok=12   changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
DC1-SVC3B                  : ok=12   changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

Thursday 30 July 2020  01:35:33 +0000 (0:00:28.279)       0:02:40.933 ********* 
=============================================================================== 
eos_validate_state ---------------------------------------------------- 160.80s
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
total ----------------------------------------------------------------- 160.80s
Thursday 30 July 2020  01:35:33 +0000 (0:00:28.279)       0:02:40.931 ********* 
=============================================================================== 
eos_validate_state : validate BGP sessions state -------------------------------------------------------------------------------------------------------------------- 28.28s
eos_validate_state : validate interfaces operational state ---------------------------------------------------------------------------------------------------------- 24.63s
eos_validate_state : validate interfaces admin state ---------------------------------------------------------------------------------------------------------------- 24.43s
eos_validate_state : validate LLDP topology ------------------------------------------------------------------------------------------------------------------------- 22.22s
eos_validate_state : validate IP reachability (directly connected) -------------------------------------------------------------------------------------------------- 17.66s
eos_validate_state : validate ArBGP is configured and operating ------------------------------------------------------------------------------------------------------ 8.28s
eos_validate_state : validate NTP status ----------------------------------------------------------------------------------------------------------------------------- 8.08s
eos_validate_state : run show version on EOS devices ----------------------------------------------------------------------------------------------------------------- 7.69s
eos_validate_state : validate reload cause --------------------------------------------------------------------------------------------------------------------------- 7.56s
eos_validate_state : Print some details regarding EOS devices -------------------------------------------------------------------------------------------------------- 5.96s
eos_validate_state : validate MLAG status ---------------------------------------------------------------------------------------------------------------------------- 5.06s
eos_validate_state : include device intended structure configuration variables --------------------------------------------------------------------------------------- 0.95s
Playbook run took 0 days, 0 hours, 2 minutes, 40 seconds
```

Physical Topology Test Results:

```
PLAY RECAP *********************************************************************************************************************************************************************************************************************
LEAF1A                     : ok=19   changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=3   
LEAF2A                     : ok=20   changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=1   
LEAF2B                     : ok=20   changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=1   
SPINE1                     : ok=19   changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1   
SPINE2                     : ok=19   changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1   

Thursday 30 July 2020  11:59:28 -0400 (0:00:06.532)       0:01:05.388 ********* 
=============================================================================== 
eos_validate_state ----------------------------------------------------- 65.33s
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
total ------------------------------------------------------------------ 65.33s
Thursday 30 July 2020  11:59:28 -0400 (0:00:06.532)       0:01:05.388 ********* 
=============================================================================== 
arista.avd.eos_validate_state : validate interfaces admin state --------------------------------------------------------------------------------------------------------------------------------------------------------- 9.16s
arista.avd.eos_validate_state : validate interfaces operational state --------------------------------------------------------------------------------------------------------------------------------------------------- 8.87s
arista.avd.eos_validate_state : validate BGP sessions state ------------------------------------------------------------------------------------------------------------------------------------------------------------- 6.53s
arista.avd.eos_validate_state : validate IP reachability (directly connected) ------------------------------------------------------------------------------------------------------------------------------------------- 5.44s
arista.avd.eos_validate_state : validate LLDP topology ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 4.31s
arista.avd.eos_validate_state : validate NTP status --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 3.28s
arista.avd.eos_validate_state : run show version on EOS devices --------------------------------------------------------------------------------------------------------------------------------------------------------- 3.23s
arista.avd.eos_validate_state : collect fan status ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- 2.61s
arista.avd.eos_validate_state : validate transceivers manufacturers ----------------------------------------------------------------------------------------------------------------------------------------------------- 2.57s
arista.avd.eos_validate_state : validate temperature -------------------------------------------------------------------------------------------------------------------------------------------------------------------- 2.41s
arista.avd.eos_validate_state : validate reload cause ------------------------------------------------------------------------------------------------------------------------------------------------------------------- 2.38s
arista.avd.eos_validate_state : collect transceivers inventory ---------------------------------------------------------------------------------------------------------------------------------------------------------- 2.27s
arista.avd.eos_validate_state : validate ArBGP is configured and operating ---------------------------------------------------------------------------------------------------------------------------------------------- 2.24s
arista.avd.eos_validate_state : collect power supplies status ----------------------------------------------------------------------------------------------------------------------------------------------------------- 2.10s
arista.avd.eos_validate_state : validate fan status (power supplies) ---------------------------------------------------------------------------------------------------------------------------------------------------- 1.76s
arista.avd.eos_validate_state : validate fan status (fan tray) ---------------------------------------------------------------------------------------------------------------------------------------------------------- 1.72s
arista.avd.eos_validate_state : validate power supplies status ---------------------------------------------------------------------------------------------------------------------------------------------------------- 1.57s
arista.avd.eos_validate_state : validate MLAG status -------------------------------------------------------------------------------------------------------------------------------------------------------------------- 1.44s
arista.avd.eos_validate_state : Print some details regarding EOS devices ------------------------------------------------------------------------------------------------------------------------------------------------ 1.26s
arista.avd.eos_validate_state : include device intended structure configuration variables ------------------------------------------------------------------------------------------------------------------------------- 0.19s
Playbook run took 0 days, 0 hours, 1 minutes, 5 seconds
```

carlbuchmann committed:
+ eos_validate_state role

carlbuchmann committed:
test desired state enable multi-agent

carlbuchmann committed:
remove test for ansible version

carlbuchmann committed:
re-order tasks + update task name

carlbuchmann reviewed (commented):
I think we should consider re-writing some of the validation tasks with an assert statement as opposed to leveraging `waitfor`, which imho provides better consistency in writing the tests.

example for lldp checks:

```
    - name: Gather LLDP topology state
      eos_command:
        commands: "show lldp neighbors {{ item.key }} | json"
      loop: "{{ ethernet_interfaces | default({}) | dict2items }}"
      when: |
        (ethernet_interfaces is defined and ethernet_interfaces is not none) and
        (item.value.peer is defined and item.value.peer is not none) and
        (item.value.peer_interface is defined and item.value.peer_interface is not none) and
        (hostvars[item.value.peer] is defined and hostvars[item.value.peer] is not none) and
        (hostvars[item.value.peer]['dns_domain'] is defined and hostvars[item.value.peer]['dns_domain'] is not none)
      ignore_errors: "{{ validation_mode_loose }}"
      register: lldp_topology
      tags:
        - avd

    - name: validate lldp topology
      assert:
        that:
          - item.stdout[0].lldpNeighbors[0] is defined
          - item.stdout[0].lldpNeighbors[0].neighborDevice == item.item.value.peer + "." + hostvars[item.item.value.peer]['dns_domain']
          - item.stdout[0].lldpNeighbors[0].neighborPort == item.item.value.peer_interface
        quiet: true
      loop: "{{ lldp_topology.results }}"
      ignore_errors: "{{ validation_mode_loose }}"
      tags:
        - avd
```

carlbuchmann committed:
typo

carlbuchmann committed:
reactor lldp test to assert method

carlbuchmann committed:
refactor ArBGP + BGP session state

carlbuchmann committed:
refactor bgp neighbors tests

carlbuchmann committed:
refactor and optimize lldp test

carlbuchmann committed:
reafactor ip reachability test

titom73 commented on ansible_collections/arista/avd/roles/eos_validate_state/tasks/main.yml:
```diff
@@ -0,0 +1,281 @@
+      tags:
+        - platform_information
+
+    - name: Display device platform and release information
```
We should use include per type of validation to make it easier to maintain:

```
- name: Load HW validation
  include_tasks: "hardware.yml"
  tags:
    - hardware
```

Also having a constraint to exclude vEOS from HW tests might help for testing

```
- name: Load HW validation
  include_tasks: "hardware.yml"
  tags:
    - hardware
  when:
    - eos_version.stdout_lines.0.modelName != 'vEOS'
```

carlbuchmann replied:
+1 agreed with this recommendation!

titom73 commented on ansible_collections/arista/avd/roles/eos_validate_state/tasks/main.yml:
```diff
@@ -0,0 +1,281 @@
+      when: ( ip_reachability_test.skipped is not defined )
+      register: ip_reachability_results
+      tags:
+        - ip_reachability
```
Maybe we could reduce number of tags by using more generic name like:

- __topology__ to cover `ip_reachability` and `lldp_topology`
- __bgp__ to cover `ArBGP_state`, `bgp_**`

carlbuchmann replied:
Agreed, what about blending this approach, as we can have multiple tags on a task. This way a group of test or specific tests can be run.

titom73 commented on ansible_collections/arista/avd/roles/eos_validate_state/tasks/main.yml:
```diff
@@ -0,0 +1,281 @@
+      when: ( service_routing_protocols_model is defined and service_routing_protocols_model == "multi-agent" )
+      ignore_errors: "{{ validation_mode_loose }}"
+      tags:
+        - ArBGP_state
```
Only lowercase should be used to reduce risk of confusion.

carlbuchmann replied:
agreed!

titom73 reviewed (changes requested):
Only reviewed tasks file

titom73 commented on ansible_collections/arista/avd/roles/eos_validate_state/tasks/main.yml:
```diff
@@ -0,0 +1,281 @@
+        interval: 2
+      ignore_errors: "{{ validation_mode_loose }}"
+      tags:
+        - optional
```
Maybe using `optional` and [`never`](https://docs.ansible.com/ansible/latest/user_guide/playbooks_tags.html#special-tags) would be more ansible standard as it is skipped by default and make it really optional with explicit definition during playbook execution

carlbuchmann replied:
yes, will add the - never tag!

titom73 committed:
Make ansible-test installation non-interactive

Commit:
move hardware validation to dedicated yml

Commit:
make task explicitly optional

Commit:
refactor interface state + new tests

titom73 committed:
Split main task file into per tag file

Add per tag file for following tags:
- avd
- bgp_check
- ip_reachability
- lldp_topology

Commit:
refactor hardware tests

Commit:
refactor ntp + mlag tests

Commit:
register results

Commit:
+ title

Commit:
+ title

Commit:
add task name

Commit:
+ register results

Commit:
standardize task names

Commit:
fix indentation

titom73 committed:
Fix YAML linting issue

titom73 commented on ansible_collections/arista/avd/roles/eos_validate_state/tasks/interface_state.yml:
```diff
@@ -0,0 +1,81 @@
+  loop_control:
+    loop_var: vlan_interface
+  ignore_errors: "{{ validation_mode_loose }}"
+  when: (vlan_interfaces is defined and vlan_interfaces is not none)
```
In case there is no SVI configured on the device, task generates an error (use-case: L2LEAF)

```
TASK [eos_validate_state : Validate Vlan interfaces state] *************
Friday 07 August 2020  11:47:16 +0200 (0:00:01.696)       0:00:19.003 *********
Friday 07 August 2020  11:47:16 +0200 (0:00:01.696)       0:00:19.003 *********
fatal: [DC1-L2LEAF1A]: FAILED! =>
  msg: dict2items requires a dictionary, got <class 'NoneType'> instead.
...ignoring
```

The same error appears for port-channel: use case is leaf not running MLAG

Trying to find a workaround

carlbuchmann replied:
Yes, I experience the same error, this is because when combining Conditionals with a loop, the when: statement is processed separately for each item, so this is why we are seeing this error... https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html

I believe we can resolve by introducing a block, and applying conditional on the block.
https://docs.ansible.com/ansible/latest/user_guide/playbooks_blocks.html

carlbuchmann replied:
after testing and reading how the conditional applies to a block, the block strategy will not work either.
"The directive affects the block itself, but is inherited by the tasks enclosed by a block. i.e. a when will be applied to the tasks, not the block itself"

titom73 replied:
Indeed. I am investigating with `set_fact` to see if we can do something to create a default empy structure for l2leafs

titom73 replied:
Only solution I found was to include tasks_file with a conditional

```yaml
- name: Build fact
  set_fact:
    temp_asset: yes
  when: (vlan_interfaces is defined and vlan_interfaces is not none and vlan_interfaces|length > 1)
  tags:
    - interfaces_state

- name: Execute SVI check
  include_tasks: "temp.yml"
  when: (temp_asset is defined and temp_asset==true)
  tags:
    - interfaces_state
```

Not optimal at all, but seems to work

carlbuchmann replied:
I was able to make it work with set_fact like this:

```yaml
- name: Set vlan_interfaces to test
  set_fact:
    vlan_interfaces_test: "{{ vlan_interfaces }}"
    cacheable: yes
  when: (vlan_interfaces is defined and vlan_interfaces is not none)
  tags:
    - interfaces_state

- name: Validate Vlan interfaces state
  assert:
    that:
      - interfaces_state.stdout[0].interfaceDescriptions[vlan_interface.key].interfaceStatus == "up"
      - interfaces_state.stdout[0].interfaceDescriptions[vlan_interface.key].lineProtocolStatus == "up"
    quiet: true
  loop: "{{ vlan_interfaces_test | default({}) | dict2items }}"
  loop_control:
    loop_var: vlan_interface
  ignore_errors: "{{ validation_mode_loose }}"
  register: vlan_interface_state_results
  tags:
    - interfaces_state
```

Note I tried first to set var to an empty dictionary but was not successful.

yes, include_task is another option but will create more separate files.

What do you think?

titom73 replied:
Your approach is far more efficient: no additional files and easy to troubleshoot.
We should apply same approach for Port Channel to cover non-mlag leaves

carlbuchmann replied:
ok, found another way with default filter set like this: ` | default({}, true)`
This sets value when variable evaluates to empty.
See commit 15790d1

Commit:
move bgp conditional to include task

Commit:
fix indentation

Commit:
remove conditionals, no longer required

Commit:
re-order test

Commit:
update register value

Commit:
initial CSV report

carlbuchmann commented:
Initial CSV report:

- [x] Hardware results
- [x] NTP
- [x] Interface State
- [x] LLDP Topology
- [x] MLAG
- [x] IP Reachability
- [x] BGP
- [x] Reload Reason

Commit:
default to dictionary when variable is empty

ksator commented:
working on hw validation and hw report

ksator committed:
fix hw tests and add hw report

ksator committed:
add fail_msg in assert tasks for ntp and hw. update csv report with fail_msg

Commit:
add fail_msg remainder of asserts

Commit:
make bgp peer tests generic

Commit:
refactor reload_cause + update template

Commit:
remove debug

Commit:
add defaults + move reports tp seperate yaml file

Commit:
+ test cEOSLab

Commit:
+ test category

Commit:
initial markdown summary report

carlbuchmann commented:
**Playbook updates:**

- [x] update playbook defaults

**Update MD Report:**
- [x] Add a summary of devices under test

**Update Role README.md**

- [x] update content to reflect latest changes
- [x] add diagram to show input and output

**Update AVD Collection README.md**

- [x] update playbook examples to include eos_validate_state
- [x] update diagrams accordingly

**BGP Check tasks:**
- [x] Run bgp session validation tests only when ArBGP test is successful.

Commit:
update ansible version to install

carlbuchmann committed:
Merge branch 'devel' into audit

Commit:
update ansible-test to ansible-2.9

carlbuchmann committed:
Merge branch 'latest devel' into audit

carlbuchmann committed:
fix liniting issue

carlbuchmann committed:
remove testing output

carlbuchmann committed:
update defaults

carlbuchmann committed:
+ duts summary

carlbuchmann committed:
Merge branch 'devel' of git://github.com/aristanetworks/ansible-avd into audit

titom73 committed:
Create folder for audit w/ eos_validate_state_dir

Create a new task in eos_validate_state_dir to support creation of
report folder.

```
root_dir: '{{ inventory_dir }}'
eos_validate_state_name: 'report'
eos_validate_state_dir: '{{root_dir}}/{{eos_state_report_name}}'
```

This variables have been added to eos_validate_state within default
section to point to reports.

titom73 commented:
Add REPORT folder part of `arista.avd.build_output_folders` role to automatically generate folder like all others

titom73 committed:
Run bgp validation tests only when ArBGP test is successful.

carlbuchmann committed:
Merge branch 'devel' of git://github.com/aristanetworks/ansible-avd into audit

carlbuchmann committed:
update role documentation

titom73 reviewed (approved):
Looks Good !

carlbuchmann committed:
fortmating

carlbuchmann committed:
update playbook examples + eos_validate_state
